lmdeploy/serve/utils.py

```diff
@@ -0,0 +1,221 @@
+                    offx = max(iter_len * i - begin, 0)
+                    offy = max(end - iter_len * (i + 1), 0)
+                    emb = emb[offx:emb.shape[0] - offy]
+                    off = max(begin - iter_len * i, 0)
+                    rng = [off, off + emb.shape[0]]
+                    iembs.append(emb)
+                    iranges.append(rng)
+
+                iembs = iembs or None
+                iranges = iranges or None
+                embs[i] = iembs
+                ranges[i] = iranges
+
+            return embs, ranges
+
+        if input_embeddings is not None:
+            if not isinstance(input_embeddings[0], list):
+                input_embeddings = [input_embeddings]
+                input_embedding_ranges = [input_embedding_ranges]
+            _input_embeddings = []
+            _input_embedding_ranges = []
+            for i in range(len(input_ids)):
+                embeddings, ranges = _split_embeddings(
+                    input_ids[i], n_max_iter, max_input_len,
+                    input_embeddings[i], input_embedding_ranges[i])
+                _input_embeddings.append(embeddings)
+                _input_embedding_ranges.append(ranges)
+            input_embeddings = _input_embeddings
+            input_embedding_ranges = _input_embedding_ranges
+
+        logits = []
+        generator = self.engine.create_instance()
+        for i in range(n_max_iter):
+            steps = [start[i] for start in index_range_starts]
+            _input_ids = [
+                input_id[start[i]:end[i]] for input_id, start, end in zip(
+                    input_ids, index_range_starts, index_range_ends)
+            ]
+            embeddings = None
+            ranges = None
+            if input_embeddings is not None:
+                embeddings = [x[i] for x in input_embeddings]
+                ranges = [x[i] for x in input_embedding_ranges]
+
+            _logits = generator.decode(_input_ids,
+                                       steps,
+                                       input_embeddings=embeddings,
+                                       input_embedding_ranges=ranges,
+                                       sequence_start=(i == 0),
+                                       sequence_end=(i == n_max_iter - 1))
+            _logits = _logits.cpu()
+            logits.append(_logits)
+
+        # concat logits. Shape is [bsz, seq_len, vocab_size]
+        logits = torch.cat(logits, dim=1)
+        return logits
+
+    def get_ppl(self, input_ids: Union[List[int], List[List[int]]]):
+        """Get perplexity scores given a list of input tokens.
+
+        Args:
+            input_ids (Union[List[int], List[List[int]]]): the batch of
+                input token ids
+        """
+        assert len(input_ids) > 0
+        if isinstance(input_ids[0], int):
+            input_ids = [input_ids]
+        for input_id in input_ids:
+            assert len(input_id) > 1
+
+        logits = self.get_logits(input_ids)
```

Comment:
The logits will occupy substantial cpu memory if the length of the prompt is long.
I think we should refer to this method: https://huggingface.co/docs/transformers/perplexity#perplexity-of-fixed-length-models